File: TChores/app/src/main/java/me/tseng/studios/tchores/java/RestaurantDetailActivity.java
package me.tseng.studios.tchores.java;

import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;

import me.tseng.studios.tchores.BuildConfig;
import me.tseng.studios.tchores.R;
import me.tseng.studios.tchores.java.adapter.RatingAdapter;
import me.tseng.studios.tchores.java.model.Rating;
import me.tseng.studios.tchores.java.model.Restaurant;
import me.tseng.studios.tchores.java.util.RestaurantUtil;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.Transaction;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import me.zhanghai.android.materialratingbar.MaterialRatingBar;

public class RestaurantDetailActivity extends AppCompatActivity
        implements EventListener<DocumentSnapshot> {

    private static final String TAG = "TChores.RestaurantDetailActivity";

    public static final String KEY_RESTAURANT_ID = BuildConfig.APPLICATION_ID + ".key_restaurant_id";   // Prefix for Intent Extra Keys
    public static final String KEY_ACTION = BuildConfig.APPLICATION_ID + ".key_action";

    public static final String ACTION_VIEW = BuildConfig.APPLICATION_ID + ".VIEW";                      // Prefix for Intent Action
    public static final String ACTION_COMPLETED = BuildConfig.APPLICATION_ID + ".COMPLETED";
    public static final String ACTION_COMPLETED_LOCALIZED = "Finished";
    public static final String ACTION_SNOOZED = BuildConfig.APPLICATION_ID + ".SNOOZED";
    public static final String ACTION_SNOOZED_LOCALIZED = "Snooze2minutes";
    public static final String ACTION_REFUSED = BuildConfig.APPLICATION_ID + ".REFUSED";
    public static final String ACTION_REFUSED_LOCALIZED = "Refuse";


    @BindView(R.id.restaurantImage)
    ImageView mImageView;

    @BindView(R.id.restaurantName)
    TextView mNameView;

    @BindView(R.id.restaurantRating)
    MaterialRatingBar mRatingIndicator;

    @BindView(R.id.restaurantNumRatings)
    TextView mNumRatingsView;

    @BindView(R.id.restaurantCity)
    TextView mCityView;

    @BindView(R.id.restaurantCategory)
    TextView mCategoryView;

    @BindView(R.id.restaurantPrice)
    TextView mPriceView;

    @BindView(R.id.viewEmptyRatings)
    ViewGroup mEmptyView;

    @BindView(R.id.recyclerRatings)
    RecyclerView mRatingsRecycler;


    private FirebaseFirestore mFirestore;
    private DocumentReference mRestaurantRef;
    private ListenerRegistration mRestaurantRegistration;
    private String mRestaurantId;

    private RatingAdapter mRatingAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_restaurant_detail);
        ButterKnife.bind(this);

        // Get restaurant ID from extras
        mRestaurantId = getIntent().getExtras().getString(KEY_RESTAURANT_ID);
        if (mRestaurantId == null) {
            throw new IllegalArgumentException("Must pass extra " + KEY_RESTAURANT_ID);
        }
        String actionId = getIntent().getExtras().getString(KEY_ACTION);
        if (actionId== null) {
            throw new IllegalArgumentException("Must pass extra " + KEY_ACTION);
        }

        Log.i(TAG, "Restaurant Detail Activity  restaurant_id=" + mRestaurantId);

        // Initialize Firestore
        mFirestore = FirebaseFirestore.getInstance();

        // Get reference to the restaurant
        mRestaurantRef = mFirestore.collection("restaurants").document(mRestaurantId);

        // Get ratings
        Query ratingsQuery = mRestaurantRef
                .collection("ratings")
                .orderBy("timestamp", Query.Direction.DESCENDING)
                .limit(50);

        // RecyclerView
        mRatingAdapter = new RatingAdapter(ratingsQuery) {
            @Override
            protected void onDataChanged() {
                if (getItemCount() == 0) {
                    mRatingsRecycler.setVisibility(View.GONE);
                    mEmptyView.setVisibility(View.VISIBLE);
                } else {
                    mRatingsRecycler.setVisibility(View.VISIBLE);
                    mEmptyView.setVisibility(View.GONE);
                }
            }
        };
        mRatingsRecycler.setLayoutManager(new LinearLayoutManager(this));
        mRatingsRecycler.setAdapter(mRatingAdapter);

    }

    @Override
    public void onStart() {
        super.onStart();

        mRatingAdapter.startListening();
        mRestaurantRegistration = mRestaurantRef.addSnapshotListener(this);
    }

    @Override
    public void onStop() {
        super.onStop();

        mRatingAdapter.stopListening();

        if (mRestaurantRegistration != null) {
            mRestaurantRegistration.remove();
            mRestaurantRegistration = null;
        }
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.slide_in_from_left, R.anim.slide_out_to_right);
    }

    /**
     * Listener for the Restaurant document ({@link #mRestaurantRef}).
     */
    @Override
    public void onEvent(DocumentSnapshot snapshot, FirebaseFirestoreException e) {
        if (e != null) {
            Log.w(TAG, "restaurant:onEvent", e);
            return;
        }

        onRestaurantLoaded(snapshot.toObject(Restaurant.class));
    }

    private void onRestaurantLoaded(Restaurant restaurant) {
        mNameView.setText(restaurant.getName());
        mRatingIndicator.setRating((float) restaurant.getAvgRating());
        mNumRatingsView.setText(getString(R.string.fmt_num_ratings, restaurant.getNumRatings()));
        mCityView.setText(restaurant.getCity());
        mCategoryView.setText(restaurant.getCategory());
        mPriceView.setText(RestaurantUtil.getPriceString(restaurant));

        // Background image
        String tempPhoto = restaurant.getPhoto();
        if (RestaurantUtil.isURL(tempPhoto)) {
            Glide.with(mImageView.getContext())
                    .load(tempPhoto)
                    .into(mImageView);
        } else {
            try {
                int tp = Integer.valueOf(tempPhoto);
                mImageView.setImageResource(tp);
            } catch (Exception e){
                // not an int or not a resource number; use default image
            }
        }

        Log.i(TAG, "Restaurant Loaded  name=" + restaurant.getName());
    }

    @OnClick(R.id.restaurantButtonBack)
    public void onBackArrowClicked(View view) {
        onBackPressed();
    }


    @OnClick(R.id.fabShowEditDialog)
    public void onEditChoreClicked(View view) {
        Intent intent = new Intent(this, RestaurantEditActivity.class);
        intent.putExtra(RestaurantEditActivity.KEY_RESTAURANT_ID, mRestaurantRef.getId());

        startActivity(intent);
    }


}

File: TChores/app/src/main/java/me/tseng/studios/tchores/kotlin/RestaurantDetailActivity.kt
package me.tseng.studios.tchores.kotlin

import android.content.Context
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.View
import android.view.inputmethod.InputMethodManager
import com.bumptech.glide.Glide
import com.google.android.gms.tasks.Task
import me.tseng.studios.tchores.R
import me.tseng.studios.tchores.kotlin.adapter.RatingAdapter
import me.tseng.studios.tchores.kotlin.model.Rating
import me.tseng.studios.tchores.kotlin.model.chore
import me.tseng.studios.tchores.kotlin.util.choreUtil
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import kotlinx.android.synthetic.main.activity_chore_detail.fabShowRatingDialog
import kotlinx.android.synthetic.main.activity_chore_detail.recyclerRatings
import kotlinx.android.synthetic.main.activity_chore_detail.choreButtonBack
import kotlinx.android.synthetic.main.activity_chore_detail.choreCategory
import kotlinx.android.synthetic.main.activity_chore_detail.choreCity
import kotlinx.android.synthetic.main.activity_chore_detail.choreImage
import kotlinx.android.synthetic.main.activity_chore_detail.choreName
import kotlinx.android.synthetic.main.activity_chore_detail.choreNumRatings
import kotlinx.android.synthetic.main.activity_chore_detail.chorePrice
import kotlinx.android.synthetic.main.activity_chore_detail.choreRating
import kotlinx.android.synthetic.main.activity_chore_detail.viewEmptyRatings

class choreDetailActivity : AppCompatActivity(),
        EventListener<DocumentSnapshot>,
        RatingDialogFragment.RatingListener {

    private var ratingDialog: RatingDialogFragment? = null

    private lateinit var firestore: FirebaseFirestore
    private lateinit var choreRef: DocumentReference
    private lateinit var ratingAdapter: RatingAdapter

    private var choreRegistration: ListenerRegistration? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chore_detail)

        // Get chore ID from extras
        val choreId = intent.extras?.getString(KEY_chore_ID)
                ?: throw IllegalArgumentException("Must pass extra $KEY_chore_ID")

        // Initialize Firestore
        firestore = FirebaseFirestore.getInstance()

        // Get reference to the chore
        choreRef = firestore.collection("chores").document(choreId)

        // Get ratings
        val ratingsQuery = choreRef
                .collection("flurrs")
                .orderBy("timestamp", Query.Direction.DESCENDING)
                .limit(50)

        // RecyclerView
        ratingAdapter = object : RatingAdapter(ratingsQuery) {
            override fun onDataChanged() {
                if (itemCount == 0) {
                    recyclerRatings.visibility = View.GONE
                    viewEmptyRatings.visibility = View.VISIBLE
                } else {
                    recyclerRatings.visibility = View.VISIBLE
                    viewEmptyRatings.visibility = View.GONE
                }
            }
        }
        recyclerRatings.layoutManager = LinearLayoutManager(this)
        recyclerRatings.adapter = ratingAdapter

        ratingDialog = RatingDialogFragment()

        choreButtonBack.setOnClickListener { onBackArrowClicked() }
        fabShowRatingDialog.setOnClickListener { onAddRatingClicked() }
    }

    public override fun onStart() {
        super.onStart()

        ratingAdapter.startListening()
        choreRegistration = choreRef.addSnapshotListener(this)
    }

    public override fun onStop() {
        super.onStop()

        ratingAdapter.stopListening()

        choreRegistration?.remove()
        choreRegistration = null
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.slide_in_from_left, R.anim.slide_out_to_right)
    }

    /**
     * Listener for the Chore document ([.choreRef]).
     */
    override fun onEvent(snapshot: DocumentSnapshot?, e: FirebaseFirestoreException?) {
        if (e != null) {
            Log.w(TAG, "chore:onEvent", e)
            return
        }

        snapshot?.let {
            val chore = snapshot.toObject(chore::class.java)
            if (chore != null) {
                onchoreLoaded(chore)
            }
        }
    }

    private fun onchoreLoaded(chore: chore) {
        choreName.text = chore.name
        choreRating.rating = chore.avgRating.toFloat()
        choreNumRatings.text = getString(R.string.fmt_num_ratings, chore.numRatings)
        choreCity.text = chore.city
        choreCategory.text = chore.category
        chorePrice.text = choreUtil.getPriceString(chore)

        // Background image
        Glide.with(choreImage.context)
                .load(chore.photo)
                .into(choreImage)
    }

    private fun onBackArrowClicked() {
        onBackPressed()
    }

    private fun onAddRatingClicked() {
        ratingDialog?.show(supportFragmentManager, RatingDialogFragment.TAG)
    }

    override fun onRating(rating: Rating) {
        // In a transaction, add the new rating and update the aggregate totals
        addRating(choreRef, rating)
                .addOnSuccessListener(this) {
                    Log.d(TAG, "Flurr added")

                    // Hide keyboard and scroll to top
                    hideKeyboard()
                    recyclerRatings.smoothScrollToPosition(0)
                }
                .addOnFailureListener(this) { e ->
                    Log.w(TAG, "Add rating failed", e)

                    // Show failure message and hide keyboard
                    hideKeyboard()
                    Snackbar.make(findViewById(android.R.id.content), "Failed to add flurr",
                            Snackbar.LENGTH_SHORT).show()
                }
    }

    private fun addRating(choreRef: DocumentReference, rating: Rating): Task<Void> {
        // Create reference for new rating, for use inside the transaction
        val ratingRef = choreRef.collection("flurrs").document()

        // In a transaction, add the new rating and update the aggregate totals
        return firestore.runTransaction { transaction ->
            val chore = transaction.get(choreRef).toObject(chore::class.java)
            if (chore == null) {
                throw Exception("Resraurant not found at ${choreRef.path}")
            }

            // Compute new number of ratings
            val newNumRatings = chore.numRatings + 1

            // Compute new average rating
            val oldRatingTotal = chore.avgRating * chore.numRatings
            val newAvgRating = (oldRatingTotal + rating.rating) / newNumRatings

            // Set new chore info
            chore.numRatings = newNumRatings
            chore.avgRating = newAvgRating

            // Commit to Firestore
            transaction.set(choreRef, chore)
            transaction.set(ratingRef, rating)

            null
        }
    }

    private fun hideKeyboard() {
        val view = currentFocus
        if (view != null) {
            (getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager)
                    .hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    companion object {

        private const val TAG = "choreDetail"

        const val KEY_chore_ID = "key_chore_id"
    }
}
